File: api/src/new_phone/schemas/sso.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class SSOProviderCreate(BaseModel):
    provider_type: str = Field(..., pattern="^(microsoft|google)$")
    display_name: str = Field(..., min_length=1, max_length=255)
    client_id: str = Field(..., min_length=1, max_length=255)
    client_secret: str = Field(..., min_length=1)
    issuer_url: str = Field(..., min_length=1, max_length=500)
    scopes: str = Field("openid email profile", max_length=500)
    auto_provision: bool = True
    default_role: str = Field("tenant_user", max_length=30)
    enforce_sso: bool = False


class SSOProviderUpdate(BaseModel):
    display_name: str | None = Field(None, min_length=1, max_length=255)
    client_id: str | None = Field(None, min_length=1, max_length=255)
    client_secret: str | None = None
    issuer_url: str | None = Field(None, min_length=1, max_length=500)
    scopes: str | None = Field(None, max_length=500)
    auto_provision: bool | None = None
    default_role: str | None = Field(None, max_length=30)
    enforce_sso: bool | None = None
    is_active: bool | None = None


class SSOProviderResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    tenant_id: uuid.UUID
    provider_type: str
    display_name: str
    client_id: str
    issuer_url: str
    discovery_url: str
    scopes: str
    auto_provision: bool
    default_role: str
    enforce_sso: bool
    is_active: bool
    created_at: datetime
    updated_at: datetime


class SSORoleMappingCreate(BaseModel):
    external_group_id: str = Field(..., min_length=1, max_length=255)
    external_group_name: str | None = Field(None, max_length=255)
    pbx_role: str = Field(..., pattern="^(tenant_admin|tenant_manager|tenant_user)$")


class SSORoleMappingResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    external_group_id: str
    external_group_name: str | None
    pbx_role: str


class SSOTestResponse(BaseModel):
    success: bool
    message: str

File: web/src/api/sso.ts
import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query"
import { apiClient } from "@/lib/api-client"
import { useAuthStore } from "@/stores/auth-store"
import { queryKeys } from "./query-keys"

// ── Types ──────────────────────────────────────────────────────────

export interface SSOCheckDomainResponse {
  sso_available: boolean
  provider_type: string | null
  display_name: string | null
  enforce_sso: boolean
}

export interface SSOInitiateResponse {
  authorization_url: string
  state: string
}

export interface SSOProviderConfig {
  id: string
  tenant_id: string
  provider_type: "microsoft" | "google"
  display_name: string
  client_id: string
  issuer_url: string
  discovery_url: string
  scopes: string
  auto_provision: boolean
  default_role: string
  enforce_sso: boolean
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface SSOProviderConfigCreate {
  provider_type: "microsoft" | "google"
  display_name: string
  client_id: string
  client_secret: string
  issuer_url: string
  scopes?: string
  auto_provision?: boolean
  default_role?: string
  enforce_sso?: boolean
}

export interface SSOProviderConfigUpdate {
  display_name?: string
  client_id?: string
  client_secret?: string
  issuer_url?: string
  scopes?: string
  auto_provision?: boolean
  default_role?: string
  enforce_sso?: boolean
  is_active?: boolean
}

export interface SSORoleMapping {
  id: string
  external_group_id: string
  external_group_name: string | null
  pbx_role: string
}

export interface SSORoleMappingCreate {
  external_group_id: string
  external_group_name?: string | null
  pbx_role: string
}

interface TokenResponse {
  access_token: string
  refresh_token: string
  token_type: string
}

// ── SSO Auth Flow ──────────────────────────────────────────────────

export function useSSOCheckDomain() {
  return useMutation({
    mutationFn: (email: string) =>
      apiClient.get<SSOCheckDomainResponse>("/api/v1/auth/sso/check-domain", { email }),
  })
}

export function useSSOInitiate() {
  return useMutation({
    mutationFn: (data: { email: string }) =>
      apiClient.post<SSOInitiateResponse>("/api/v1/auth/sso/initiate", data),
  })
}

export function useSSOComplete() {
  return useMutation({
    mutationFn: (data: { state: string }) =>
      apiClient.post<TokenResponse>("/api/v1/auth/sso/complete", data),
  })
}

// ── SSO Config CRUD ──────────────────────────────────────────────

export function useSSOConfig() {
  const tenantId = useAuthStore((s) => s.activeTenantId)!
  return useQuery({
    queryKey: queryKeys.sso.config(tenantId),
    queryFn: () => apiClient.get<SSOProviderConfig>("/api/v1/sso-config"),
    enabled: !!tenantId,
    retry: false,
  })
}

export function useCreateSSOConfig() {
  const tenantId = useAuthStore((s) => s.activeTenantId)!
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: SSOProviderConfigCreate) =>
      apiClient.post<SSOProviderConfig>("/api/v1/sso-config", data),
    onSuccess: () => qc.invalidateQueries({ queryKey: queryKeys.sso.config(tenantId) }),
  })
}

export function useUpdateSSOConfig() {
  const tenantId = useAuthStore((s) => s.activeTenantId)!
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: SSOProviderConfigUpdate) =>
      apiClient.patch<SSOProviderConfig>("/api/v1/sso-config", data),
    onSuccess: () => qc.invalidateQueries({ queryKey: queryKeys.sso.config(tenantId) }),
  })
}

export function useDeleteSSOConfig() {
  const tenantId = useAuthStore((s) => s.activeTenantId)!
  const qc = useQueryClient()
  return useMutation({
    mutationFn: () => apiClient.delete("/api/v1/sso-config"),
    onSuccess: () => qc.invalidateQueries({ queryKey: queryKeys.sso.config(tenantId) }),
  })
}

export function useTestSSOConfig() {
  return useMutation({
    mutationFn: () => apiClient.post<{ success: boolean; message: string }>("/api/v1/sso-config/test"),
  })
}

// ── Role Mappings ──────────────────────────────────────────────────

export function useSSORoleMappings() {
  const tenantId = useAuthStore((s) => s.activeTenantId)!
  return useQuery({
    queryKey: queryKeys.sso.roleMappings(tenantId),
    queryFn: () => apiClient.get<SSORoleMapping[]>("/api/v1/sso-config/role-mappings"),
    enabled: !!tenantId,
  })
}

export function useAddRoleMapping() {
  const tenantId = useAuthStore((s) => s.activeTenantId)!
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: SSORoleMappingCreate) =>
      apiClient.post<SSORoleMapping>("/api/v1/sso-config/role-mappings", data),
    onSuccess: () => qc.invalidateQueries({ queryKey: queryKeys.sso.roleMappings(tenantId) }),
  })
}

export function useDeleteRoleMapping() {
  const tenantId = useAuthStore((s) => s.activeTenantId)!
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (mappingId: string) =>
      apiClient.delete(`/api/v1/sso-config/role-mappings/${mappingId}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: queryKeys.sso.roleMappings(tenantId) }),
  })
}
